### ISSUE TITLE
Compilation with dealII fails in Teuchos_DefaultMpiComm.hpp

### ISSUE THREAD

**Issue by anates** [trilinos/Trilinos#1481]
When compiling, I get the error messages 

> In file included from /usr/local/include/Teuchos_Assert.hpp:46:0,
                 from /usr/local/include/Teuchos_any.hpp:51,
                 from /usr/local/include/Teuchos_ParameterEntry.hpp:51,
                 from /usr/local/include/Teuchos_ParameterList.hpp:51,
                 from ~/Downloads/dealii/include/deal.II/lac/trilinos_precondition.h:39,
                 from ~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:17:
/usr/local/include/Teuchos_DefaultMpiComm.hpp: In instantiation of ‘Teuchos::MpiComm<Ordinal>::MpiComm(MPI_Comm) [with Ordinal = int; MPI_Comm = ompi_communicator_t*]’:
/usr/local/include/Ifpack2_LocalFilter_def.hpp:110:20:   required from ‘Ifpack2::LocalFilter<MatrixType>::LocalFilter(const Teuchos::RCP<const Tpetra::RowMatrix<typename MatrixType::scalar_type, typename MatrixType::local_ordinal_type, typename MatrixType::global_ordinal_type, typename MatrixType::node_type> >&) [with MatrixType = Tpetra::RowMatrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >; typename MatrixType::node_type = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP>; typename MatrixType::global_ordinal_type = int; typename MatrixType::local_ordinal_type = int; typename MatrixType::scalar_type = double]’
/usr/local/include/Ifpack2_Details_DenseSolver_def.hpp:245:23:   required from ‘void Ifpack2::Details::DenseSolver<MatrixType, false>::initialize() [with MatrixType = Tpetra::RowMatrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >]’
~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:302:1:   required from here
/usr/local/include/Teuchos_DefaultMpiComm.hpp:835:3: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
   TEUCHOS_TEST_FOR_EXCEPTION(rawMpiComm == MPI_COMM_NULL,
   ^
[ 61%] Building CXX object source/numerics/CMakeFiles/obj_numerics_debug.dir/matrix_tools_once.cc.o
In file included from /usr/local/include/Teuchos_DefaultComm.hpp:48:0,
                 from /usr/local/include/MueLu.hpp:58,
                 from ~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:34:
/usr/local/include/Teuchos_DefaultMpiComm.hpp: In instantiation of ‘Teuchos::RCP<Teuchos::Comm<OrdinalType> > Teuchos::MpiComm<Ordinal>::duplicate() const [with Ordinal = int]’:
~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:302:1:   required from here
/usr/local/include/Teuchos_DefaultMpiComm.hpp:1637:12: error: invalid conversion from ‘int’ to ‘MPI_Comm {aka ompi_communicator_t*}’ [-fpermissive]
   MPI_Comm newRawComm = MPI_COMM_NULL;
            ^~~~~~~~~~
/usr/local/include/Teuchos_DefaultMpiComm.hpp: In instantiation of ‘Teuchos::RCP<Teuchos::Comm<OrdinalType> > Teuchos::MpiComm<Ordinal>::split(int, int) const [with Ordinal = int]’:
~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:302:1:   required from here
/usr/local/include/Teuchos_DefaultMpiComm.hpp:1673:15: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
   if (newComm == MPI_COMM_NULL) {
               ^
/usr/local/include/Teuchos_DefaultMpiComm.hpp: In instantiation of ‘Teuchos::RCP<Teuchos::Comm<OrdinalType> > Teuchos::MpiComm<Ordinal>::createSubcommunicator(const Teuchos::ArrayView<const int>&) const [with Ordinal = int]’:
~/Downloads/dealii/source/lac/trilinos_precondition_muelu.cc:302:1:   required from here
/usr/local/include/Teuchos_DefaultMpiComm.hpp:1741:15: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
   if (newComm == MPI_COMM_NULL) {

I am using OpenMPI with gcc7.1.1, latest versions of dealII and trilinos.

**Comment by ibaned** [comment]
This seems similar to #1480. Have you coordinated with @bangerth ?

**Comment by anates** [comment]
He suggested the post, after I introduced the fixes suggested in his issue (and got the error messages posted above)

**Comment by ibaned** [comment]
Okay, so if I understand this correctly, even after #1480 is fixed the above errors persist?

**Comment by bangerth** [comment]
Apparently the compiler thinks that `MPI_COMM_NULL` is not a pointer type (as `MPI_Comm` is). @anates -- can you try to find what type `MPI_COMM_NULL` has in the MPI installation you are using?

**Comment by mhoemmen** [comment]
@anates Could you please also enclose error messages in double backquotes, in order to avoid inadvertent formatting?
```
Here is an example.  Notice that _underscores_ and ~~strikethrough~~ markings lack their usual effect.
```
Thanks!

**Comment by anates** [comment]
Reinstall to another location in a separate folder fixed the problem apparently, the compilation of deal.II works. I assume an include-file clash, due to having multiple libraries in `/usr/local/include`
 Still used fix proposed in https://github.com/trilinos/Trilinos/issues/1480, no errors yet. Should I thus assume the issue closed?

**Comment by ibaned** [comment]
Yes, this looks a lot like the same root cause as #1480, i.e. two MPI installs fighting. `int` is the type of `MPI_Comm` in MPICH-like variants, and in OpenMPI it is a pointer. Changing `"mpi.h"` to `<mpi.h>` seems fine to me, but if two MPI variants are installed *at the same location*, I'd say thats an unsupported use case.

**Comment by nmhamster** [comment]
Yeah @ibaned - these two cannot coexist in the same install location today. There are a few projects that allow something like this if anyone needs it.

**Comment by mhoemmen** [comment]
So what's the action item here?  Do we need to change all the `"mpi.h"` to `<mpi.h>`?

**Comment by bangerth** [comment]
Yes, that's exactly what needs to happen.

**Comment by mhoemmen** [comment]
Thanks @bangerth ! :-)

I will close this issue and mark it as a duplicate of #1480.  All further discussion will take place there.
